### PR TITLE
QA-356: Makedata and Checkdata testing for Computed Values

### PR DESCRIPTION
**Pull request by MDAlFattah**
### Add checkdata and makedata test for Computed Values

There is a need to add Computed Values to RTA tests.
We used this test plan scenario as a draft for the scenario: [https://www.arangodb.com/docs/stable/data-modeling-documents-computed-values.html]()
We created files 060_computed_value.json and 060_computed_values.js to add tests in this test suite. Note that this feature was added in the 3.10 release: [https://www.arangodb.com/docs/stable/data-modeling-documents-computed-values.html]()

- [x] :pizza: New feature

### Checklist
* checked and tested manually on different deployments.

### Related Information
QA task: https://arangodb.atlassian.net/browse/QA-356
### RTA PR
https://github.com/arangodb/release-test-automation/pull/369

### CODE REVIEW

File: test_data/makedata_suites/060_computed_values.js [hunk 1]
```diff
@@ -0,0 +1,776 @@
...
+    `for doc in ${view[1]} filter doc.cv_field == to_hex(doc.name) collect with count into c return c`,
+    `for doc in ${view[1]} filter doc.cv_field == CONCAT('42_', TO_STRING(doc.field)) collect with count into c return c`,
+    `for doc in ${view[1]} search doc.cv_field1=='foo' and doc.cv_field2=='bar' and doc.cv_field3=='baz' collect with count into c return c`,
+    `for doc in ${view[1]} filter doc.cv_field == CONCAT(doc._key, ' ', doc._id, ' ', doc._rev) collect with count into c return c`
+  ];
+
+  // Expected output array for the TestView views query output
+  let views_exp_output = [64000, 64000, 64000, 0, 0, 64000, 96000, 64000, 64000, 64000, 64000, 64000, 0, 0, 64000, 64000, 96000, 64000, 64000];
+
+  return [views_array, views_exp_output];
```
Comment (alexbakharew) [inline test_data/makedata_suites/060_computed_values.js:113]: Lets store array of tuples in following way:
[
[<query_1>, <expected_output_1>],
[<query_2>, <expected_output_2>],
.....
]

File: test_data/makedata_suites/060_computed_values.js [hunk 2]
```diff
@@ -0,0 +1,776 @@
...
+
+    makeDataDB: function (options, isCluster, isEnterprise, database, dbCount) {
+      // All items created must contain dbCount
+      print(`060: making per database data ${dbCount}`);
+      print("060: Creating computed values with sample collections");
+
+      // getting all the collection names
+      let c = collection_declaration(dbCount);
+
+      let a1 = createCollectionSafe(c[0], 3, 3, { computedValues: [{ "name": "default", "expression": "RETURN SOUNDEX('sky')", overwrite: true }] });
```
Comment (alexbakharew) [inline test_data/makedata_suites/060_computed_values.js:133]: Please rename `a` to `c`. And use same number as you use for array `c`.
Should be:

```
let c0 = createCollectionSafe(c[0], 3, 3, { computedValues: [{ "name": "default", "expression": "RETURN SOUNDEX('sky')", overwrite: true }] });
```

And so on

File: test_data/makedata_suites/060_computed_values.js [hunk 3]
```diff
@@ -0,0 +1,776 @@
...
+          }
+        } 
+        else if (col === a3) {
+          if (expected_field[0].cv_field_insert !== null) {
+          } else {
+            throw new Error(`Computed value field missing from collection`);
+          }
+        }
+        else if (col === a4 || col === a5) {
+          if (expected_field[2].cv_field !== null) {
```
Comment (alexbakharew) [inline test_data/makedata_suites/060_computed_values.js:701]:
```suggestion
          if (expected_field[2].cv_field == null) {
```

Just compare it with null. And if it is null - then you go under this condition and throw a error.

Please fix the same issue in code below and above

File: test_data/makedata_suites/060_computed_values.js (hunk 1 above)
Comment (MDAlFattah) [reply test_data/makedata_suites/060_computed_values.js]: Array to tuple conversion done.

File: test_data/makedata_suites/060_computed_values.js (hunk 3 above)
Comment (MDAlFattah) [reply test_data/makedata_suites/060_computed_values.js]: Done.

**Comment by dothebart** [comment]
the cleanup is missing from this.

File: test_data/makedata_suites/060_computed_values.js [hunk 4]
```diff
@@ -0,0 +1,766 @@
...
+  ];
+
+  return views_Name_array;
+};
+
+// This method will take db and a tuple parameter containing one query and one expected output
+// as a touple elements and then compare both's results
+let resultComparision = (db, tuple) =>{
+  for (let i = 0; i < tuple.length; i++) {
+    for (let j = 0; j < tuple[i].length; j++) {
```
[the remark this answers is not in the document]
Comment (MDAlFattah) [reply test_data/makedata_suites/060_computed_values.js:39]: Done.

File: test_data/makedata_suites/060_computed_values.js [hunk 5]
```diff
@@ -0,0 +1,766 @@
...
+      var output = db._query(tuple[i][j]).toArray();
+      var newOuput = Number(output);
+      progress(``);
+      if (newOuput !== tuple[i][j+1]) {
+        throw new Error(`${tuple[i][j]} Query's output: ${newOuput} didn't match with ecxpected_output: ${tuple[i][j+1]}`);
```
[the remark this answers is not in the document]
Comment (MDAlFattah) [reply test_data/makedata_suites/060_computed_values.js:44]: Done.

File: test_data/makedata_suites/060_computed_values.js [hunk 6]
```diff
@@ -0,0 +1,766 @@
...
+    [`for doc in ${c[8]} OPTIONS { indexHint : 'persistent' } filter doc.cv_field == CONCAT('42_', TO_STRING(doc.field)) collect with count into c return c`, 64000],
+    [`for doc in ${c[9]} OPTIONS { indexHint : 'persistent' } filter doc.cv_field1 == 'foo' and doc.cv_field2 == 'bar' and doc.cv_field3 == 'baz' collect with count into c return c`, 64000],
+    [`for doc in ${c[10]} OPTIONS { indexHint : 'inverted', forceIndexHint: true, waitForSync: true } filter doc.cv_field == FIRST(for d in ${c[10]} limit 1001, 1 return CONCAT(d._key, ' ', d._id, ' ', d._rev)) collect with count into c return c`, 1],
+    [`for doc in ${c[10]} OPTIONS { indexHint : 'persistent' } filter doc.cv_field == CONCAT(doc._key, ' ', doc._id, ' ', doc._rev) collect with count into c return c`, 64000]
+  ];
+  return indexTuple;
+}
+
+// this function will provide all the queries for views
+function viewsArray(dbCount) {
```
[the remark this answers is not in the document]
Comment (MDAlFattah) [reply test_data/makedata_suites/060_computed_values.js]: Done.

File: test_data/makedata_suites/060_computed_values.js [hunk 7]
```diff
@@ -0,0 +1,766 @@
...
+
+      if (j == 0) {
+        break;
+      }
+    }
+  }
+}
+
+//execute queries which use indexes and verify that the proper amount of docs are returned
+function indexArray(dbCount){
```
[the remark this answers is not in the document]
Comment (MDAlFattah) [reply test_data/makedata_suites/060_computed_values.js]: Done.

File: test_data/makedata_suites/060_computed_values.js [hunk 8]
```diff
@@ -0,0 +1,766 @@
...
+
+      let c12_actual_modification = c12.properties({ computedValues: [{ "name": "cv_field", "expression": "RETURN [{from_doc: CONCAT(@doc.name, ' ', @doc.field), system:{_key: @doc._key, _rev: @doc._rev, _id: @doc._id}, values: [RANGE(1, 10)]}]", "overwrite": true }] });
+
+      checkComValProperties(c[11], c12_exp_modification, c12_actual_modification.computedValues);
+
+      //-------------------------------------------------------x-------------------------------------------------------------
+
+      c1.ensureIndex({"type":"inverted","name":"inverted","fields":[{"name":"cv_field"}]});
+      setTimeout(function() {
+        c1.ensureIndex({"type":"persistent","name":"persistent","fields":["cv_field"], "sparse": true});
```
[the remark this answers is not in the document]
Comment (MDAlFattah) [reply test_data/makedata_suites/060_computed_values.js]: Done.

File: test_data/makedata_suites/060_computed_values.js [hunk 9]
```diff
@@ -0,0 +1,766 @@
...
+      print(`060: checking data ${dbCount}`);
+
+      // get all the view's variable name from the variable_name_declaration method globally
+      let view = views_name_declaration(dbCount);
+
+      try {
+        db._dropView(`${view[0]}`);
+        db._dropView(`${view[1]}`);
+      } catch (e) {
+        print(e);
```
[the remark this answers is not in the document]
Comment (MDAlFattah) [reply test_data/makedata_suites/060_computed_values.js:734]: Done.

File: test_data/makedata_suites/060_computed_values.js [hunk 10]
```diff
@@ -0,0 +1,766 @@
...
+};
+
+// This method will take db and a tuple parameter containing one query and one expected output
+// as a touple elements and then compare both's results
+let resultComparision = (db, tuple) =>{
+  for (let i = 0; i < tuple.length; i++) {
+    for (let j = 0; j < tuple[i].length; j++) {
+      var output = db._query(tuple[i][j]).toArray();
+      var newOuput = Number(output);
+      progress(``);
```
[the remark this answers is not in the document]
Comment (MDAlFattah) [reply test_data/makedata_suites/060_computed_values.js]: Done.

**Review by alexbakharew — APPROVED** [review]
LGTM

File: test_data/makedata_suites/101_collections.js [hunk 11]
```diff
@@ -0,0 +1,700 @@
...
+    `c8_overwriteTrue_101_${dbCount}`,
+    `c9_multiple_101_${dbCount}`,
+    `c10_101_${dbCount}`
+  ];
+};
+
+// this method will declare all the views name with proper dbCount
+let views_names_declaration = (dbCount) =>{
+  return [
+    `test_view_${dbCount}`,
```
Comment (dothebart) [inline test_data/makedata_suites/101_collections.js:23]: please add `_101` here too

File: test_data/makedata_suites/101_collections.js [hunk 12]
```diff
@@ -0,0 +1,700 @@
...
+// This method will take db and a tuple parameter containing one query and one expected output
+// as a tuple elements and then compare both's results
+let result_comparison = (db, tuple) =>{
+  for (let i = 0; i < tuple.length; i++) {
+    let query_str = tuple[i][0];
+    let expected_output =  tuple[i][1];
+    var output = db._query(query_str).toArray();
+    var newOuput = Number(output);
+    if (newOuput !== expected_output) {
+      throw new Error(`${query_str} query_str's output: ${newOuput} didn't match with ecxpected_output: ${expected_output}`);
```
Comment (dothebart) [inline test_data/makedata_suites/101_collections.js:37]: please add `101: `

File: test_data/makedata_suites/101_collections.js [hunk 13]
```diff
@@ -0,0 +1,700 @@
...
+// this function will compare properties objects
+function compareProperties(name, obj1, obj2) {
+  if (obj1.hasOwnProperty('optimizeTopK')) {
+    delete obj1.optimizeTopK;
+  }
+  if (obj2.hasOwnProperty('optimizeTopK')) {
+    delete obj2.optimizeTopK;
+  }
+  if(_.isEqual(obj1, obj2) === false){
+    throw new Error(`Properties missmatched for the collection/view ${name} ${JSON.stringify(obj1)} <-> ${JSON.stringify(obj2)}`);
```
Comment (dothebart) [inline test_data/makedata_suites/101_collections.js:122]: 101:

**Review by dothebart — APPROVED** [review]
please add the remaining prefixes, else LGTM.

File: test_data/makedata_suites/101_collections.js (hunk 12 above)
Comment (MDAlFattah) [reply test_data/makedata_suites/101_collections.js:37]: Done.

File: test_data/makedata_suites/101_collections.js (hunk 13 above)
Comment (MDAlFattah) [reply test_data/makedata_suites/101_collections.js:122]: Done.

**Pull request merged**